evalml/pipelines/multiseries_regression_pipeline.py

```diff
@@ -163,16 +163,45 @@ def predict_in_sample(
             objective,
             calculating_residuals,
         )
+        unstacked_predictions = unstacked_predictions[
+            [
+                series_id_target
+                for series_id_target in y_train_unstacked.columns
+                if series_id_target in unstacked_predictions.columns
+            ]
+        ]
+        unstacked_predictions.index = X_unstacked[self.time_index]
         stacked_predictions = stack_data(
             unstacked_predictions,
-            include_series_id=include_series_id,
+            include_series_id=True,
             series_id_name=self.series_id,
         )
-
+        stacked_predictions = stacked_predictions.reset_index()
```

Comment:
The goal of this snippet is to set the index as the time index column, stack the data (and thus using the dates in the time index column to generate new stacked dates) and then resetting the index so that the resulting time index column can be used when we `pd.merge` later on in line 193. 

While it's possible to just copy over the time_index column from `X` after stacking, I think it's safer to just generate it from the `X_unstacked` index like this as we know for sure that the X_unstacked time_index aligns  with the unstacked_predictions whereas it's technically possible to have an `X` time_index that's out of order (and thus would be incorrect if we simply copied over this column). I'm open to suggestions for a cleaner implementation!

Reply:
Ok, I think I understand now! I think a comment would be great. I also wonder if it would be useful to explicitly say `reset_index(drop=False)`, so that even if pandas changes their defaults we don't get screwed. 

My motivation here is that this is something that might be confusing to someone looking back at it in the future, since the goal isn't clear from the code itself. I hope that makes sense!

Reply:
Good call on the reset index parameter, I'll add that in. I'll add a clarifying comment or two so that it's clear what's going on here.

Your motivation makes sense! I feel like I've been so lost in the weeds of this implementation for a while now so it's good to have multiple pairs of eyes on this to highlight what's intuitive and what isn't 😅